### PR TITLE
Add more tests for JWT allow_missing_or_failed

### PR DESCRIPTION
**Pull request by diemtvu**
Description:
Add more unit tests to demonstrate the status quo behavior of JWT filter with `allow_missing_or_failed` requirement. Some behavior is arguably incorrect or confusing, though we will try not to fix it in this PR.

Related issues:
- https://github.com/envoyproxy/envoy/issues/8909

Risk Level: Low
Adding tests only.
Testing:
Add more unit tests for existing feature.
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by lizan** [comment]
can you fix format and DCO?

File: test/extensions/filters/http/jwt_authn/all_verifier_test.cc [hunk 1]
```diff
@@ -18,16 +18,75 @@ namespace HttpFilters {
...
+    forward_payload_header: "x-other-payload"
+    local_jwks:
+      inline_string: ""
+rules:
+- match:
+    path: "/"
+)";
+
+const std::string kExampleHeader{"x-example"};
+const std::string kOtherHeader{"x-other"};
```
Comment (lizan) [inline test/extensions/filters/http/jwt_authn/all_verifier_test.cc]: no static or global std::string, (use `constexpr absl::string_view` instead.

File: test/extensions/filters/http/jwt_authn/all_verifier_test.cc [hunk 2]
```diff
@@ -18,16 +18,75 @@ namespace HttpFilters {
...
+
+// Returns true if the jwt_header is empty, and the jwt_header payload exists.
+// Based on the JWT provider setup for this test, this matcher is equivalent to JWT verification
+// was success.
+MATCHER_P(JwtOutputSuccess, jwt_header, "") {
+  auto payload_header = std::string(jwt_header) + "-payload";
```
Comment (lizan) [inline test/extensions/filters/http/jwt_authn/all_verifier_test.cc]: nit: absl::StrCat

File: test/extensions/filters/http/jwt_authn/all_verifier_test.cc (hunk 1 above)
Comment (diemtvu) [reply test/extensions/filters/http/jwt_authn/all_verifier_test.cc]: Changed to constexpr char[]. string_view need to do few castings later.

File: test/extensions/filters/http/jwt_authn/all_verifier_test.cc (hunk 2 above)
Comment (diemtvu) [reply test/extensions/filters/http/jwt_authn/all_verifier_test.cc]: Done

File: test/extensions/filters/http/jwt_authn/all_verifier_test.cc (hunk 1 above)
Comment (lizan) [reply test/extensions/filters/http/jwt_authn/all_verifier_test.cc]: either is fine :)

**Comment by diemtvu** [comment]
> can you fix format and DCO?

Done

**Comment by diemtvu** [comment]
@lizan , can you re-approve? I think I lost it after a push.

**Pull request merged**